botnim/kb/download_sources.py

```diff
@@ -5,51 +5,85 @@
 from io import StringIO
 import yaml
 from ..config import get_logger
+from typing import List, Tuple, BinaryIO
+import io
 
 logger = get_logger(__name__)
 
-def download_and_convert_spreadsheet(source_url: str, target_dir: Path, context_name: str) -> None:
-    """Download and convert Google Spreadsheet data to individual markdown files"""
+def download_and_convert_spreadsheet(source_url: str, target_dir: Path, context_name: str, force_all: bool = True) -> List[Tuple[str, BinaryIO, str]]:
+    """Download and convert Google Spreadsheet data to memory buffers and cache files
+    
+    Args:
+        source_url: URL of the Google Spreadsheet
+        target_dir: Directory to store cache files
+        context_name: Name of the context for file naming
+        force_all: If True, return all entries regardless of changes
+        
+    Returns:
+        List of tuples (filename, file_buffer, content_type) for entries
+    """
     try:
+        logger.info(f"Downloading spreadsheet from {source_url} to {target_dir}")
         sheet_id = source_url.split('/d/')[1].split('/')[0]
         url = f'https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=csv'
         
         response = requests.get(url)
-        response.encoding = 'utf-8'  # Explicitly set response encoding
+        response.encoding = 'utf-8'
         response.raise_for_status()
         
-        print("Raw response text:")
-        print(response.text[:500])  # Print first 500 chars to see the structure
+        # Load existing files for comparison
```

Comment:
Why do we need to download to actual files?

Reply:
To support local actions (like exploring chunking, or prompt fine tunes). They can also act as a cache to detect changes in the source spreadsheet.

Reply:
Not sure how creating local files helps chunking or fine tunes (as you're going to load it to your local ES anyway...), but either way see my other comment on caching.

Reply:
was added by mistake directly in the repository.

Reply:
to be fixed.